e2e/pages/any.page.ts

```diff
@@ -43,54 +17,113 @@ export class AnyPage {
-                    .all(by.xpath(buttonPath))
-                    .isPresent();
-            },
-            Wait.normal,
-            'Button or link did not show in time'
-        );
+        const button = await element
+            .all(by.xpath(buttonPath))
+            .filter(e => e.isPresent() && e.isDisplayed())
+            .first();
+
+        if (await button.isPresent()) {
+            await button.click();
+            return;
+        }
 
-        const buttonElement =
-            await element
-                .all(by.xpath(buttonPath))
-                .first();
+        const anchorPath = '//a[normalize-space()="' + linkText + '"]';
 
-        return await buttonElement.isEnabled();
+        const anchor = await element
+            .all(by.xpath(anchorPath))
+            .filter(e => e.isPresent() && e.isDisplayed())
+            .first();
+
+        if (await anchor.isPresent()) {
+            await anchor.click();
+            return;
+        }
+
+        const linkPath =
+            '//*[self::label or self::span]' +
+            '[text()[normalize-space()="' + linkText + '"]]';
+
+        const link = await element
+            .all(by.xpath(linkPath))
+            .filter(e => e.isPresent() && e.isDisplayed())
+            .first();
+
+        await link.click();
     }
 
-    async linkContains(match: string) {
+    async isButtonEnabled(match: string, shortWait: boolean = false) {
 
         try {
 
-            const linkPath = '//*[self::button or self::a][contains(normalize-space(), "' + match + '")]';
+            await browser.wait(
+                async () => {
+                    return (await element
+                        .all(by.xpath(
+                            '//*[self::button or self::a]' +
+                            '[contains(normalize-space(), "' + match + '") and not(ancestor::*[@hidden])]'
+                        ))
+                        .filter(e => e.isPresent() && e.isDisplayed() && e.isEnabled())
+                        .count()) > 0;
+                },
+                shortWait ? Wait.short : Wait.normal
+            );
 
-            return await element(by.xpath(linkPath)).isDisplayed()
-                && await element(by.xpath(linkPath)).isEnabled();
+            return true;
 
         } catch (error) {
             return false;
         }
+
+        // const buttonPath = '//*[self::button or self::a][normalize-space()="' + buttonText + '"]';
+        //
+        // await browser.wait(
+        //     async () => {
+        //         return await element
+        //             .all(by.xpath(buttonPath))
+        //             .isPresent();
+        //     },
+        //     Wait.normal,
+        //     'Button or link did not show in time'
+        // );
+        //
+        // const buttonElement =
+        //     await element
+        //         .all(by.xpath(buttonPath))
+        //         .first();
+        //
+        // return await buttonElement.isEnabled();
```

Comment:
Intentionally left in?

Reply:
No -- but I did remove in a subsequent PR :)